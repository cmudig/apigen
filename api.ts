import * as ts from "typescript";
import * as fs from 'fs';
import { ASTStatement } from "./internalRepresentation"
import { emitter, capitalize, decapitalize } from "./util"
import { generateExportFunction, generateClass, generateClassMemberMethod, MethodArg, createArgString } from "./generate";
// import { fieldDefs } from "vega-lite/build/src/encoding";
// import { encoding } from "./scratch/classes";

const emit = emitter('')
let fieldDefArgs: MethodArg[] = [];

export function generateVLAPI(statement: ASTStatement) {

  const className = capitalize(statement.name as string);

  switch (statement.kind) {
    case ts.SyntaxKind.TypeAliasDeclaration: {

      if (statement.name == "ColorDef") {
        break;
      }

      if (statement.name == "PositionDef") {
        break;
      }

      if (statement.name == "ValueDef") {
        break;
      }

      let argName = "";
      if (statement.type == ts.SyntaxKind.UnionType) {
        argName = "type";
      }
      else {
        argName = decapitalize(className);
      }
      //prepare constructor argument string
      let argString: MethodArg[] = [];
      if (statement.args != undefined) {
        argString = [{ keyword: "private", name: argName, type: `"${statement.args.join('" | "')}"` }];
      }
      if (statement.name == "Spec") {
        argString = [{ keyword: "private", name: "mark", type: "Mark" }, { keyword: "private", name: "encode", type: "Encoding" }]
      }
      emit(generateClass(className, createArgString(argString, true), []));
      emit(generateExportFunction(className, className, argString));
      break;
    }

    case ts.SyntaxKind.InterfaceDeclaration: {

      if (statement.name == "LayerSpec") {
        break;
      }
      if (statement.name == "FieldDef") {

        for (let [name, type] of Object.entries(statement.members)) {
          fieldDefArgs.push({ keyword: "private", name: name, type: statement.memberUnionTypes[name] ? `"${statement.memberUnionTypes[name].join(`" | "`)}"` : "string" });
        }
        break;
      }

      if (statement.name == "Encoding") {

        let encodingArgs: MethodArg[] = [];

        for (let [name, type] of Object.entries(statement.members)) {

          emit(generateClass(capitalize(name), createArgString(fieldDefArgs, true), []));

          emit(generateExportFunction(name, name, fieldDefArgs));
          encodingArgs.push({ keyword: "private", name: `${name}?`, type: `${capitalize(name)}` });
        }


        emit(generateClass(statement.name, createArgString(encodingArgs, true), []));

        emit(generateExportFunction("encode", statement.name, encodingArgs));
        break;
      }
    }
  }
}

export function generatetoSpec() {
  emit(`export function toSpec(obj: any){
      return JSON.stringify(obj);
    }`);
}

//Write to the output file.
export function writeFile(outputFile: string) {
  fs.writeFileSync(outputFile, emit.code());
}

export function getEmitCode(): string {
    return emit.code();
}