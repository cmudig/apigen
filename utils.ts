const Data = Symbol('data');
const Context = Symbol('context');
let id_counter = 0;

export function id(prefix: any) {
  return (prefix || '') + (++id_counter);
}

// export class BaseObject {
//   toObject() { return toObject(this); }
// }

// export class MergeObject {
//   constructor(values: any) { this[Data] = values; }
//   toObject() { return Object.assign({}, ...recurse(this[Data])); }
// }

// export function assign(target: any, ...sources: any) {
//   if (sources.length === 1 && Array.isArray(sources[0])) {
//     target[Data] = sources[0];
//   } else {
//     sources.forEach(s: any => {
//       Object.assign(target[Data], isObject(s) && s[Data] || s)
//     });
//   }
//   return target;
// }

export function flat(value: any) {
  return Array.isArray(value) ? [].concat(...value) : value;
}

export function prop(obj: any, name: any) {
  return isObject(obj) && !obj[Data] ? obj[name] : get(obj, name);
}

export function get(obj: any, name: any) {
  return obj[Data][name];
}

export function set(obj: any, name: any, value: any) {
  obj[Data][name] = object(value);
}

function duplicate(...obj: any) {
  return Object.assign(
    Object.create(Object.getPrototypeOf(obj[0])),
    ...obj
  );
}

export function copy(obj: any) {
  return duplicate(obj, { [Data]: Object.assign({}, obj[Data]) });
}

export function init(obj: any, value: any) {
  obj[Data] = value || {};
}

export function annotate(value: any, context: any) {
  const ctx = { [Context]: context };
  return isArray(value)
    ? value.map(v => duplicate(v, ctx))
    : duplicate(value, ctx)
}

// function recurse(d: any) {
//   return d && d.toObject ? d.toObject(d[Context] || 0) : toObject(d);
// }

function toObject(value: any) {
  if (isArray(value)) {
    return value.map(d => recurse(d));
  } else if (isObject(value)) {
    const data = value[Data] || value;
    return isArray(data)
      ? recurse(data)
      : Object.keys(data).reduce((_, k) => {
          _[k] = recurse(data[k]);
          return _;
        }, {});
  } else {
    return value;
  }
}

export function raw(value: any) {
  return { [Data]: value, toObject: () => value };
}

function object(value: any) {
  return (isObject(value) && !value[Data]) ? {[Data]: value || {}} : value;
}

// export function merge(flag: any, ...values: any) {
//   const objects = [].concat(...values);
//   return new MergeObject(flag ? annotate(objects, flag) : objects);
// }

// export function nest(obj: any, keys: any, rest: any) {
//   const m = keys.reduce((m, k) => (m[k] = 1, m), {}),
//         u = {}, v = {};

//   for (let k in obj) (m[k] ? u : v)[k] = obj[k];
//   u[rest] = v;
//   return u;
// }

// -- type checkers --

export const isArray = Array.isArray;

export function isBoolean(_: any) {
  return typeof _ === 'boolean';
}

export function isIterable(_: any) {
  return isObject(_) && typeof _[Symbol.iterator] === 'function';
}

export function isNumber(_: any) {
  return typeof _ === 'number';
}

export function isObject(_: any) {
  return _ === Object(_) && !isArray(_);
}

export function isString(_: any) {
  return typeof _ === 'string';
}

export function isFunction(_: any) {
  return typeof _ === 'function';
}

export function isEventTarget(_: any) {
  return isObject(_) && isFunction(_.addEventListener);
}
