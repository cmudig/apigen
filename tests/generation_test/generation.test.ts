import { parseFile, validateOutput } from "../utils";
import { generateVLAPI, generatetoSpec, getEmitCode } from "../../api"
import * as fs from 'fs';
import exp from "constants";

/**
 * Test the generated file is the same as expected.
 */
describe("VLType test", () => {

    it("Output should be same as expected", () => {
        const input_file_path = "tests/generation_test/test_files";
        const expected_file_path = "tests/generation_test/expected_generated_files";
        // const input_file = `${input_file_path}/enumType.ts`
        // const expected_file = `${expected_file_path}/enumType.txt`
        validateOutput(input_file_path + "/complexType.ts", expected_file_path + "/complexType.txt");

        // TODO: clear the emit code and support multiple files
        // validateOutput("tests/generation_test/test_files/complexType.ts", "tests/generation_test/expected_generated_files/complexType.txt");
    });

});