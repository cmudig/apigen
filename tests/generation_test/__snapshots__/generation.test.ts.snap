// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VLType test Output should be same as expected 1`] = `
"
class Mark {
    constructor(private type: "bar" | "area" | "line") {}
    
  }
  
export function mark(type: "bar" | "area" | "line"){
    return new Mark(type);
  }

class X {
    constructor(private field: string, private type: "quantitative" | "ordinal") {}
    
  }
  
export function x(field: string, type: "quantitative" | "ordinal"){
    return new X(field, type);
  }

class Y {
    constructor(private field: string, private type: "quantitative" | "ordinal") {}
    
  }
  
export function y(field: string, type: "quantitative" | "ordinal"){
    return new Y(field, type);
  }

class Color {
    constructor(private field: string, private type: "quantitative" | "ordinal") {}
    
  }
  
export function color(field: string, type: "quantitative" | "ordinal"){
    return new Color(field, type);
  }

class Encoding {
    constructor(private x?: X, private y?: Y, private color?: Color) {}
    
  }
  
export function encode(x?: X, y?: Y, color?: Color){
    return new Encoding(x, y, color);
  }

class Spec {
    constructor(private mark: Mark, private encode: Encoding) {}
    
  }
  
export function spec(mark: Mark, encode: Encoding){
    return new Spec(mark, encode);
  }

export function toSpec(obj: any){
      return JSON.stringify(obj);
    }"
`;
